### ISSUE TITLE
Further enhancement of facilities to provide hyperparameters

### ISSUE THREAD

**Issue by dkmisra** [microsoft/Intrepid#5]
We should further improve the hyperparameter facility to provide nested hyperparameters, e.g., we should be able to more transparently change the learning rate of a particular module while keeping it the same for others. Perhaps using Hydra might help (https://github.com/pytorch/hydra-torch).

**Comment by dkmisra** [comment]
Ideally, we want the following:

- When running an experiment, the method provides an explanation of what each hyperparameter does
- If a provided hyperparameter value isn't used then this is marked and brought to the attention
- Provide a way to specify hyperparameters for different modules (maybe nested structures will be good here). Let's say we have two models that we are using called cnn1 and cnn3. Each of them is using 4 hyperparameters. Then it should be possible for me to change them using command line arguments separately. However, if we don't set the module's hyperparameter, it should use a default value or inherit the hyperparameter from a parent class.
